### PR TITLE
feat(Yim Menu): Added multiple featues and small fixes

### PR DESCRIPTION
**Pull request by xM4ddy**
Updated Weapons.h, made welcome message include current open key, added clean player and keep player clean, invisibility, force crosshairs, no recoil, no spread, get all weapons, and weapon damage multiplier.

### CODE REVIEW

**Comment by xM4ddy** [comment]
#64 Addresses some of these

File: BigBaseV2/src/backend/backend.cpp [hunk 1]
```diff
@@ -41,6 +41,8 @@ namespace big
 			looped::self_police();
 			looped::self_no_ragdoll();
 			looped::self_super_run();
+			looped::self_invisibility();
+			looped::self_clean_player();
```
Comment (Yimura) [inline BigBaseV2/src/backend/backend.cpp]: Set these alphabetically, to follow the same coding style as the entire project.

File: BigBaseV2/src/backend/looped/looped.hpp [hunk 2]
```diff
@@ -17,10 +17,12 @@ namespace big
...
 		static void self_free_cam();
 		static void self_godmode();
-		static void self_off_radar();
-		static void self_police();
+		static void self_invisibility();
+		static void self_clean_player();
 		static void self_noclip();
 		static void self_no_ragdoll();
+		static void self_off_radar();
+		static void self_police();
```
Comment (Yimura) [inline BigBaseV2/src/backend/looped/looped.hpp]: Set these alphabetically, to follow the same coding style as the entire project.

File: BigBaseV2/src/backend/looped/weapons/no_recoil.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
+#include "pointers.hpp"
+#include "CPedFactory.hpp"
```
Comment (Yimura) [inline BigBaseV2/src/backend/looped/weapons/no_recoil.cpp]: No need to include CPedFactory.hpp, it is already included by common.hpp.

File: BigBaseV2/src/backend/looped/weapons/no_recoil.cpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+#include "backend/looped/looped.hpp"
```
Comment (Yimura) [inline BigBaseV2/src/backend/looped/weapons/no_recoil.cpp]: Set these alphabetically, to follow the same coding style as the entire project.

File: BigBaseV2/src/backend/looped/weapons/no_recoil.cpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+	float get_og_recoil_value(uint32_t hash)
+	{
+		return std::find_if(og_recoil_values.begin(), og_recoil_values.end(), [hash](auto const entry)
+			{
+				return hash == entry.first;
+			})->second;
+	}
+
+	auto get_recoil_value(uint32_t hash)
```
Comment (Yimura) [inline BigBaseV2/src/backend/looped/weapons/no_recoil.cpp]: You use float as a return type in the function above this but then use auto here, stay consistent pick one, not both.

File: BigBaseV2/src/backend/looped/weapons/no_spread.cpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	//			return hash == entry.first;
+	//		})->second;
+	//}
+
+	//auto get_spread_value(uint32_t hash)
+	//{
+	//	return g->weapons.no_spread
+	//		? 0.f
+	//		: get_og_spread_value(hash);
+	//}
```
Comment (Yimura) [inline BigBaseV2/src/backend/looped/weapons/no_spread.cpp]: Remove comments, unless this code is intended for future use.

File: BigBaseV2/src/backend/looped/weapons/no_spread.cpp [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+			//auto const cur_weapon_hash = weapon_mgr->m_selected_weapon_hash;
+			//if (prev_weapon_hash != cur_weapon_hash)
+			//{
+			//	if (!is_spread_value_cached(cur_weapon_hash))
+			//	{
+			//		og_spread_values.push_back({ cur_weapon_hash, weapon_mgr->m_weapon_info->m_spread });
+			//	}
+
+			//	weapon_mgr->m_weapon_info->m_spread = 0.f; //get_spread_value(cur_weapon_hash);
+			//}
```
Comment (Yimura) [inline BigBaseV2/src/backend/looped/weapons/no_spread.cpp]: Remove comments, unless this code is intended for future use.

File: BigBaseV2/src/core/globals.hpp [hunk 8]
```diff
@@ -64,6 +64,8 @@ namespace big
 			};
 
 			bool godmode = false;
+			bool invisibility = false;
+			bool clean_player = false;
```
Comment (Yimura) [inline BigBaseV2/src/core/globals.hpp]: Set these alphabetically, to follow the same coding style as the entire project.

File: BigBaseV2/src/backend/looped/weapons/no_spread.cpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
+#include "backend/looped/looped.hpp"
+#include "pointers.hpp"
+#include "CPedFactory.hpp"
```
Comment (Yimura) [inline BigBaseV2/src/backend/looped/weapons/no_spread.cpp]: No need to include CPedFactory.hpp, it is already included by common.hpp.

File: BigBaseV2/src/core/globals.hpp [hunk 10]
```diff
@@ -196,6 +201,8 @@ namespace big
 			this->tunables.no_idle_kick = j["tunables"]["no_idle_kick"];
 
 			this->self.godmode = j["self"]["godmode"];
+			this->self.invisibility = j["self"]["invisibility"];
+			this->self.invisibility = j["self"]["clean_player"];
```
Comment (Yimura) [inline BigBaseV2/src/core/globals.hpp]: ~~Assigning 2 different json values to the same global?~~

Set these alphabetically, to follow the same coding style as the entire project.

File: BigBaseV2/src/core/globals.hpp [hunk 11]
```diff
@@ -284,6 +294,8 @@ namespace big
 				{
 					"self", {
 						{ "godmode", this->self.godmode },
+						{ "invisibility", this->self.invisibility },
+						{ "clean_player", this->self.clean_player },
```
Comment (Yimura) [inline BigBaseV2/src/core/globals.hpp]: Set these alphabetically, to follow the same coding style as the entire project.

File: BigBaseV2/src/views/view_self.cpp [hunk 12]
```diff
@@ -31,6 +31,24 @@ namespace big
...
+			ImGui::Checkbox("Invisibility", &g->self.invisibility);
+			ImGui::Checkbox("Keep Player Clean", &g->self.clean_player);
+			if (ImGui::Button("Clean Player"))
+			{
+				QUEUE_JOB_BEGIN_CLAUSE()
+				{
+					PED::CLEAR_PED_BLOOD_DAMAGE(PLAYER::PLAYER_PED_ID());
+					PED::CLEAR_PED_WETNESS(PLAYER::PLAYER_PED_ID());
+					PED::CLEAR_PED_ENV_DIRT(PLAYER::PLAYER_PED_ID());
+					PED::RESET_PED_VISIBLE_DAMAGE(PLAYER::PLAYER_PED_ID());
```
Comment (Yimura) [inline BigBaseV2/src/views/view_self.cpp]: Make these calls a function, you're using these same natives again in a looped feature.

util/entity.hpp
```cpp
inline void clean_ped(Ped ped)
{
	PED::CLEAR_PED_BLOOD_DAMAGE(PLAYER::PLAYER_PED_ID());
	PED::CLEAR_PED_WETNESS(PLAYER::PLAYER_PED_ID());
	PED::CLEAR_PED_ENV_DIRT(PLAYER::PLAYER_PED_ID());
	PED::RESET_PED_VISIBLE_DAMAGE(PLAYER::PLAYER_PED_ID());
}
```

File: BigBaseV2/src/backend/looped/self/clean_player.cpp [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+{
+	void looped::self_clean_player()
+	{
+		auto const player_ped = PLAYER::PLAYER_PED_ID();
+
+		if (g->self.clean_player) {
+			PED::CLEAR_PED_BLOOD_DAMAGE(player_ped);
+			PED::CLEAR_PED_WETNESS(player_ped);
+			PED::CLEAR_PED_ENV_DIRT(player_ped);
+			PED::RESET_PED_VISIBLE_DAMAGE(player_ped);
```
Comment (Yimura) [inline BigBaseV2/src/backend/looped/self/clean_player.cpp]: Make these calls a function, you're using these same natives again in a looped feature.

util/entity.hpp
```cpp
inline void clean_ped(Ped ped)
{
	PED::CLEAR_PED_BLOOD_DAMAGE(PLAYER::PLAYER_PED_ID());
	PED::CLEAR_PED_WETNESS(PLAYER::PLAYER_PED_ID());
	PED::CLEAR_PED_ENV_DIRT(PLAYER::PLAYER_PED_ID());
	PED::RESET_PED_VISIBLE_DAMAGE(PLAYER::PLAYER_PED_ID());
}
```

**Review by Yimura — CHANGES_REQUESTED** [review]

File: BigBaseV2/src/backend/looped/weapons/no_recoil.cpp [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+	{
+		return g->weapons.no_recoil
+			? 0.f
+			: get_og_recoil_value(hash);
+	}
+
+	void looped::weapons_no_recoil()
+	{
+		auto* const weapon_mgr = (*g_pointers->m_ped_factory)->m_local_ped->m_weapon_manager;
```
Comment (Yimura) [inline BigBaseV2/src/backend/looped/weapons/no_recoil.cpp]: Why access weapon_mgr like this?
There's a global which exposes access to the local player ped.

```cpp
auto* const weapon_mgr = g_local_player->m_weapon_manager;
```

Comment (Yimura) [reply BigBaseV2/src/backend/looped/weapons/no_recoil.cpp]: Additionally check if the g_local_player != nullptr, this is a pain point in both yours and my code that might crash you if you inject early and the CPed pointer isn't accessible.

File: BigBaseV2/src/backend/looped/weapons/no_spread.cpp [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+	//auto get_spread_value(uint32_t hash)
+	//{
+	//	return g->weapons.no_spread
+	//		? 0.f
+	//		: get_og_spread_value(hash);
+	//}
+
+	void looped::weapons_no_spread()
+	{
+		auto* const weapon_mgr = (*g_pointers->m_ped_factory)->m_local_ped->m_weapon_manager;
```
Comment (Yimura) [inline BigBaseV2/src/backend/looped/weapons/no_spread.cpp]: Why access weapon_mgr like this?
There's a global which exposes access to the local player ped.

```cpp
auto* const weapon_mgr = g_local_player->m_weapon_manager;
```

Comment (Yimura) [reply BigBaseV2/src/backend/looped/weapons/no_spread.cpp]: Additionally check if the g_local_player != nullptr, this is a pain point in both yours and my code that might crash you if you inject early and the CPed pointer isn't accessible.

File: BigBaseV2/src/backend/looped/weapons/no_spread.cpp [hunk 16]
```diff
@@ -1,51 +1,50 @@
-#include "backend/looped/looped.hpp"
 #include "pointers.hpp"
 #include "CPedFactory.hpp"
+#include "backend/looped/looped.hpp"
```
Comment (Yimura) [inline BigBaseV2/src/backend/looped/weapons/no_spread.cpp]: Set these alphabetically, to follow the same coding style as the entire project.

No need to include CPedFactory.hpp, it's already included automatically by common.hpp

File: BigBaseV2/src/gui.cpp [hunk 17]
```diff
@@ -90,7 +92,7 @@ namespace big
 
 	void gui::script_init()
 	{
-		g_notification_service->push("Welcome", "Loaded YimMenu. Press INSERT to open");
+		g_notification_service->push("Welcome", std::string("Loaded YimMenu. Press ").append(ImGui::key_names[g->settings.hotkeys.menu_toggle]).append(" to open"));
```
Comment (Yimura) [inline BigBaseV2/src/gui.cpp]: Use fmt::format instead of string appending.

File: BigBaseV2/src/views/view_weapons.cpp [hunk 18]
```diff
@@ -1,5 +1,9 @@
 #include "core/data/custom_weapons.hpp"
 #include "views/view.hpp"
+#include "script.hpp"
+#include "gta/Weapons.h"
+
+#include "fiber_pool.hpp"
```
Comment (Yimura) [inline BigBaseV2/src/views/view_weapons.cpp:2]: Set these alphabetically, to follow the same coding style as the entire project.

File: BigBaseV2/src/views/view_weapons.cpp [hunk 19]
```diff
@@ -27,6 +31,25 @@ namespace big
...
+				}
+				QUEUE_JOB_END_CLAUSE
+			}
+
+			float damage_slider;
+			damage_slider = g->weapons.increased_damage;
+			if (ImGui::SliderFloat("Damage Multiplier", &damage_slider, 1.f, 10.f, "%.1f"))
+			{
+				g->weapons.increased_damage = damage_slider;
+			}
```
Comment (Yimura) [inline BigBaseV2/src/views/view_weapons.cpp]: Why aren't you directly assigning to g->weapons.increased_damage?
This is unnecessary.

**Review by Yimura — CHANGES_REQUESTED** [review]

File: BigBaseV2/src/backend/looped/weapons/no_spread.cpp (hunk 6 above)
Comment (Yimura) [reply BigBaseV2/src/backend/looped/weapons/no_spread.cpp]: Seems this code is used in future commits.

File: BigBaseV2/src/backend/looped/weapons/no_spread.cpp (hunk 7 above)
Comment (Yimura) [reply BigBaseV2/src/backend/looped/weapons/no_spread.cpp]: Seems this code is used in future commits.

File: BigBaseV2/src/backend/looped/weapons/no_recoil.cpp (hunk 3 above)
Comment (Yimura) [reply BigBaseV2/src/backend/looped/weapons/no_recoil.cpp]: I'm incorrect, but still remove CPedFactory.hpp, look at one of the comments further down, you won't need the CPedFactory you can directly access the local CPed.

File: BigBaseV2/src/backend/looped/weapons/no_spread.cpp (hunk 9 above)
Comment (Yimura) [reply BigBaseV2/src/backend/looped/weapons/no_spread.cpp]: I'm incorrect, but still remove CPedFactory.hpp, look at one of the comments further down, you won't need the CPedFactory you can directly access the local CPed.

**Comment by xM4ddy** [comment]
All requested changes should be fixed.

**Comment by Yimura** [comment]
Alright thank you for fixing those things, I will test them later today.

**Comment by xM4ddy** [comment]
So that's how you do that.

**Comment by Yimura** [comment]
Anyways tell me if you're done with this PR, drafts is meant for if you're still planning to push.

**Review by Yimura — APPROVED** [review]

**Comment by Yimura** [comment]
Code looks good, I will squash these commits and merge them into master.

**Pull request merged**